Error when running remax with multiturn rollouts

In the latest version of verl I can successfully run multiturn training with grpo:

https://github.com/volcengine/verl/blob/main/examples/sglang_multiturn/run_qwen2.5-3b_gsm8k_multiturn_4xgpu.sh

But if I switch the advantage estimator from grpo to remax I get the following error message:

`  File "/root/verl/verl/trainer/main_ppo.py", line 282, in run
    trainer.fit()
  File "/root/verl/verl/trainer/ppo/ray_trainer.py", line 1169, in fit
    batch = batch.union(gen_batch_output)
  File "/root/verl/verl/protocol.py", line 685, in union
    self.non_tensor_batch = union_numpy_dict(self.non_tensor_batch, other.non_tensor_batch)
  File "/root/verl/verl/protocol.py", line 190, in union_numpy_dict
    assert _deep_equal(tensor_dict1[key], tensor_dict2[key], visited=set()), (
AssertionError: `messages` in tensor_dict1 and tensor_dict2 are not the same object.`

--- a/verl/experimental/agent_loop/agent_loop.py
+++ b/verl/experimental/agent_loop/agent_loop.py
@@ -433,6 +433,7 @@ class AgentLoopWorker:
             response_mask: | 1, 1, 1, ..., 1, 1 | 0, 0, .., 0, 0 | 1, 1, 1, ..., 1, 1 | 0, 0, ..., 0|
         """
         config = self.rollout_config
+        validate = batch.meta_info.get("validate", False)
         sampling_params = dict(
             temperature=config.temperature,
             top_p=config.top_p,
@@ -441,8 +442,13 @@ class AgentLoopWorker:
             logprobs=config.calculate_log_probs,
         )
 
+        def apply_greedy_sampling_params(params: dict[str, Any]) -> None:
+            params["top_p"] = 1.0
+            params["top_k"] = -1
+            params["temperature"] = 0
+
         # override sampling params for validation
-        if batch.meta_info.get("validate", False):
+        if validate:
             sampling_params["top_p"] = config.val_kwargs.top_p
             sampling_params["top_k"] = config.val_kwargs.top_k
             sampling_params["temperature"] = config.val_kwargs.temperature
@@ -477,13 +483,19 @@ class AgentLoopWorker:
             batch.meta_info.get("global_steps", -1), index.tolist(), batch.meta_info.get("validate", False)
         )
 
+        # NOTE: __do_sample__ is an internal per-sample override used by REMAX combined rollout.
+        # Do not forward it to concrete agent loops, which may reject unknown kwargs.
+        per_sample_do_sample = batch.non_tensor_batch.get("__do_sample__")
         tasks = []
         for i in range(len(batch)):
             trace_this_sample = i in traced_indices
-            kwargs = {k: v[i] for k, v in batch.non_tensor_batch.items()}
+            kwargs = {k: v[i] for k, v in batch.non_tensor_batch.items() if k != "__do_sample__"}
+            sample_sampling_params = dict(sampling_params)
+            if not validate and per_sample_do_sample is not None and not bool(per_sample_do_sample[i]):
+                apply_greedy_sampling_params(sample_sampling_params)
             tasks.append(
                 asyncio.create_task(
-                    self._run_agent_loop(sampling_params, trajectory_info[i], trace=trace_this_sample, **kwargs)
+                    self._run_agent_loop(sample_sampling_params, trajectory_info[i], trace=trace_this_sample, **kwargs)
                 )
             )
         outputs = await asyncio.gather(*tasks)

--- a/verl/experimental/separation/ray_trainer.py
+++ b/verl/experimental/separation/ray_trainer.py
@@ -20,7 +20,6 @@ This trainer supports model-agonistic model initialization with huggingface
 """
 
 import uuid
-from copy import deepcopy
 from pprint import pprint
 from typing import Any, Optional
 

--- a/verl/experimental/separation/ray_trainer.py
+++ b/verl/experimental/separation/ray_trainer.py
@@ -437,22 +437,21 @@ class SeparateRayPPOTrainer(RayPPOTrainer):
             gen_batch_output.pop(non_tensor_batch_keys=["__do_sample__"])
 
         if self.config.algorithm.adv_estimator == AdvantageEstimator.REMAX:
-            with marked_timer("gen_max", timing_raw, color="purple"):
-                gen_baseline_output = combined_gen_output.slice(num_sampled_prompts, None)
-                if "__do_sample__" in gen_baseline_output.non_tensor_batch:
-                    gen_baseline_output.pop(non_tensor_batch_keys=["__do_sample__"])
+            gen_baseline_output = combined_gen_output.slice(num_sampled_prompts, None)
+            if "__do_sample__" in gen_baseline_output.non_tensor_batch:
+                gen_baseline_output.pop(non_tensor_batch_keys=["__do_sample__"])
 
-                # REMAX only needs one scalar baseline reward per original prompt.
-                # Agent-loop rollout outputs contain sample-specific non-tensor fields
-                # such as turns, tool rewards and extras; keep the baseline path isolated.
-                if self.use_rm and "rm_scores" not in gen_baseline_output.batch.keys():
-                    baseline_reward = self._compute_reward_colocate(gen_baseline_output)
-                    gen_baseline_output = gen_baseline_output.union(baseline_reward)
+            # REMAX only needs one scalar baseline reward per original prompt.
+            # Agent-loop rollout outputs contain sample-specific non-tensor fields
+            # such as turns, tool rewards and extras; keep the baseline path isolated.
+            if self.use_rm and "rm_scores" not in gen_baseline_output.batch.keys():
+                baseline_reward = self._compute_reward_colocate(gen_baseline_output)
+                gen_baseline_output = gen_baseline_output.union(baseline_reward)
 
-                reward_baseline_tensor = gen_baseline_output.batch["rm_scores"].sum(dim=-1)
-                batch.batch["reward_baselines"] = reward_baseline_tensor
+            reward_baseline_tensor = gen_baseline_output.batch["rm_scores"].sum(dim=-1)
+            batch.batch["reward_baselines"] = reward_baseline_tensor
 
-                del gen_baseline_output
+            del gen_baseline_output
         del combined_gen_batch, combined_gen_output
         # repeat to align with repeated responses in rollout
         batch = batch.repeat(repeat_times=self.config.actor_rollout_ref.rollout.n, interleave=True)

--- a/verl/experimental/separation/ray_trainer.py
+++ b/verl/experimental/separation/ray_trainer.py
@@ -404,47 +404,56 @@ class SeparateRayPPOTrainer(RayPPOTrainer):
         gen_batch = self._get_gen_batch(batch)
         # pass global_steps to trace
         gen_batch.meta_info["global_steps"] = self.global_steps
-        gen_batch_output = gen_batch.repeat(repeat_times=self.config.actor_rollout_ref.rollout.n, interleave=True)
+        rollout_n = self.config.actor_rollout_ref.rollout.n
+        gen_batch_output = gen_batch.repeat(repeat_times=rollout_n, interleave=True)
+
+        if self.config.algorithm.adv_estimator == AdvantageEstimator.REMAX:
+            # NOTE: REMAX needs one sampled rollout plus one greedy baseline per prompt.
+            # Keep them in a single agent-loop/vLLM request to avoid sending a second
+            # rollout after replicas have been put to sleep, which can leave async vLLM
+            # engines in an invalid state for multi-turn agent workloads.
+            gen_batch_output.non_tensor_batch["__do_sample__"] = np.ones(len(gen_batch_output), dtype=bool)
+            gen_baseline_batch = deepcopy(gen_batch)
+            gen_baseline_batch.non_tensor_batch["__do_sample__"] = np.zeros(len(gen_baseline_batch), dtype=bool)
+            combined_gen_batch = DataProto.concat([gen_batch_output, gen_baseline_batch])
+            num_sampled_prompts = len(gen_batch_output)
+        else:
+            combined_gen_batch = gen_batch_output
+            num_sampled_prompts = len(gen_batch_output)
 
         with marked_timer("gen", timing_raw, color="red"):
             if self.curr_step_profile:
                 self.async_rollout_manager.start_profile(global_step=self.global_steps)
-            gen_batch_output = self.async_rollout_manager.generate_sequences(gen_batch_output)
+            combined_gen_output = self.async_rollout_manager.generate_sequences(combined_gen_batch)
             self.checkpoint_manager.sleep_replicas()
             if self.curr_step_profile:
                 self.llm_server_manager.stop_profile()
 
-            timing_raw.update(gen_batch_output.meta_info["timing"])
-            gen_batch_output.meta_info.pop("timing", None)
+            timing_raw.update(combined_gen_output.meta_info["timing"])
+            combined_gen_output.meta_info.pop("timing", None)
+
+        gen_batch_output = combined_gen_output.slice(0, num_sampled_prompts)
+        if "__do_sample__" in gen_batch_output.non_tensor_batch:
+            gen_batch_output.pop(non_tensor_batch_keys=["__do_sample__"])
 
         if self.config.algorithm.adv_estimator == AdvantageEstimator.REMAX:
             with marked_timer("gen_max", timing_raw, color="purple"):
-                gen_baseline_batch = deepcopy(gen_batch)
-                gen_baseline_batch.meta_info["do_sample"] = False
-                if self.curr_step_profile:
-                    self.llm_server_manager.start_profile()
-                gen_baseline_output = self.async_rollout_manager.generate_sequences(gen_baseline_batch)
-                self.checkpoint_manager.sleep_replicas()
-                if self.curr_step_profile:
-                    self.llm_server_manager.stop_profile()
-                batch = batch.union(gen_baseline_output)
-                # compute reward model score on batch
-                rm_scores = None
-                if self.use_rm and "rm_scores" not in batch.batch.keys():
-                    batch_reward = self._compute_reward_colocate(batch)
-                    batch = batch.union(batch_reward)
+                gen_baseline_output = combined_gen_output.slice(num_sampled_prompts, None)
+                if "__do_sample__" in gen_baseline_output.non_tensor_batch:
+                    gen_baseline_output.pop(non_tensor_batch_keys=["__do_sample__"])
 
-                # Compute or extract reward for REMAX baseline
-                reward_baseline_tensor = batch.batch["rm_scores"].sum(dim=-1)
+                # REMAX only needs one scalar baseline reward per original prompt.
+                # Agent-loop rollout outputs contain sample-specific non-tensor fields
+                # such as turns, tool rewards and extras; keep the baseline path isolated.
+                if self.use_rm and "rm_scores" not in gen_baseline_output.batch.keys():
+                    baseline_reward = self._compute_reward_colocate(gen_baseline_output)
+                    gen_baseline_output = gen_baseline_output.union(baseline_reward)
 
-                keys_to_pop = set(gen_baseline_output.batch.keys())
-                if rm_scores is not None:
-                    keys_to_pop.update(rm_scores.batch.keys())
-                batch.pop(batch_keys=list(keys_to_pop))
-
+                reward_baseline_tensor = gen_baseline_output.batch["rm_scores"].sum(dim=-1)
                 batch.batch["reward_baselines"] = reward_baseline_tensor
 
-                del rm_scores, gen_baseline_batch, gen_baseline_output
+                del gen_baseline_output
+        del combined_gen_batch, combined_gen_output
         # repeat to align with repeated responses in rollout
         batch = batch.repeat(repeat_times=self.config.actor_rollout_ref.rollout.n, interleave=True)
         batch = batch.union(gen_batch_output)

--- a/verl/experimental/separation/ray_trainer.py
+++ b/verl/experimental/separation/ray_trainer.py
@@ -413,7 +413,7 @@ class SeparateRayPPOTrainer(RayPPOTrainer):
             # rollout after replicas have been put to sleep, which can leave async vLLM
             # engines in an invalid state for multi-turn agent workloads.
             gen_batch_output.non_tensor_batch["__do_sample__"] = np.ones(len(gen_batch_output), dtype=bool)
-            gen_baseline_batch = deepcopy(gen_batch)
+            gen_baseline_batch = gen_batch.slice(0, None)
             gen_baseline_batch.non_tensor_batch["__do_sample__"] = np.zeros(len(gen_baseline_batch), dtype=bool)
             combined_gen_batch = DataProto.concat([gen_batch_output, gen_baseline_batch])
             num_sampled_prompts = len(gen_batch_output)

--- a/verl/trainer/ppo/ray_trainer.py
+++ b/verl/trainer/ppo/ray_trainer.py
@@ -1353,9 +1353,22 @@ class RayPPOTrainer:
 
                 # pass global_steps to trace
                 gen_batch.meta_info["global_steps"] = self.global_steps
-                gen_batch_output = gen_batch.repeat(
-                    repeat_times=self.config.actor_rollout_ref.rollout.n, interleave=True
-                )
+                rollout_n = self.config.actor_rollout_ref.rollout.n
+                gen_batch_output = gen_batch.repeat(repeat_times=rollout_n, interleave=True)
+
+                if self.config.algorithm.adv_estimator == AdvantageEstimator.REMAX:
+                    # NOTE: REMAX needs one sampled rollout plus one greedy baseline per prompt.
+                    # Keep them in a single agent-loop/vLLM request to avoid sending a second
+                    # rollout after replicas have been put to sleep, which can leave async vLLM
+                    # engines in an invalid state for multi-turn agent workloads.
+                    gen_batch_output.non_tensor_batch["__do_sample__"] = np.ones(len(gen_batch_output), dtype=bool)
+                    gen_baseline_batch = deepcopy(gen_batch)
+                    gen_baseline_batch.non_tensor_batch["__do_sample__"] = np.zeros(len(gen_baseline_batch), dtype=bool)
+                    combined_gen_batch = DataProto.concat([gen_batch_output, gen_baseline_batch])
+                    num_sampled_prompts = len(gen_batch_output)
+                else:
+                    combined_gen_batch = gen_batch_output
+                    num_sampled_prompts = len(gen_batch_output)
 
                 is_last_step = self.global_steps >= self.total_training_steps
                 with marked_timer("step", timing_raw):
@@ -1363,42 +1376,32 @@ class RayPPOTrainer:
                     with marked_timer("gen", timing_raw, color="red"):
                         if curr_step_profile:
                             self.llm_server_manager.start_profile()
-                        gen_batch_output = self.async_rollout_manager.generate_sequences(gen_batch_output)
+                        combined_gen_output = self.async_rollout_manager.generate_sequences(combined_gen_batch)
                         self.checkpoint_manager.sleep_replicas()
                         if curr_step_profile:
                             self.llm_server_manager.stop_profile()
 
-                        timing_raw.update(gen_batch_output.meta_info["timing"])
-                        gen_batch_output.meta_info.pop("timing", None)
+                        timing_raw.update(combined_gen_output.meta_info["timing"])
+                        combined_gen_output.meta_info.pop("timing", None)
+
+                    gen_batch_output = combined_gen_output.slice(0, num_sampled_prompts)
+                    if "__do_sample__" in gen_batch_output.non_tensor_batch:
+                        gen_batch_output.pop(non_tensor_batch_keys=["__do_sample__"])
 
                     if self.config.algorithm.adv_estimator == AdvantageEstimator.REMAX:
-                        with marked_timer("gen_max", timing_raw, color="purple"):
-                            gen_baseline_batch = deepcopy(gen_batch)
-                            gen_baseline_batch.meta_info["do_sample"] = False
-                            if curr_step_profile:
-                                self.llm_server_manager.start_profile()
-                            gen_baseline_output = self.async_rollout_manager.generate_sequences(gen_baseline_batch)
-                            self.checkpoint_manager.sleep_replicas()
-                            if curr_step_profile:
-                                self.llm_server_manager.stop_profile()
-                            batch = batch.union(gen_baseline_output)
-                            # compute reward model score on batch
-                            rm_scores = None
-                            if self.use_rm and "rm_scores" not in batch.batch.keys():
-                                batch_reward = self._compute_reward_colocate(batch)
-                                batch = batch.union(batch_reward)
+                        gen_baseline_output = combined_gen_output.slice(num_sampled_prompts, None)
+                        if "__do_sample__" in gen_baseline_output.non_tensor_batch:
+                            gen_baseline_output.pop(non_tensor_batch_keys=["__do_sample__"])
 
-                            # Compute or extract reward for REMAX baseline
-                            reward_baseline_tensor = batch.batch["rm_scores"].sum(dim=-1)
+                        if self.use_rm and "rm_scores" not in gen_baseline_output.batch.keys():
+                            baseline_reward = self._compute_reward_colocate(gen_baseline_output)
+                            gen_baseline_output = gen_baseline_output.union(baseline_reward)
 
-                            keys_to_pop = set(gen_baseline_output.batch.keys())
-                            if rm_scores is not None:
-                                keys_to_pop.update(rm_scores.batch.keys())
-                            batch.pop(batch_keys=list(keys_to_pop))
+                        reward_baseline_tensor = gen_baseline_output.batch["rm_scores"].sum(dim=-1)
+                        batch.batch["reward_baselines"] = reward_baseline_tensor
 
-                            batch.batch["reward_baselines"] = reward_baseline_tensor
-
-                            del rm_scores, gen_baseline_batch, gen_baseline_output
+                        del gen_baseline_output
+                    del combined_gen_batch, combined_gen_output
                     # repeat to align with repeated responses in rollout
                     batch = batch.repeat(repeat_times=self.config.actor_rollout_ref.rollout.n, interleave=True)
                     batch = batch.union(gen_batch_output)

--- a/verl/trainer/ppo/ray_trainer.py
+++ b/verl/trainer/ppo/ray_trainer.py
@@ -1362,7 +1362,7 @@ class RayPPOTrainer:
                     # rollout after replicas have been put to sleep, which can leave async vLLM
                     # engines in an invalid state for multi-turn agent workloads.
                     gen_batch_output.non_tensor_batch["__do_sample__"] = np.ones(len(gen_batch_output), dtype=bool)
-                    gen_baseline_batch = deepcopy(gen_batch)
+                    gen_baseline_batch = gen_batch.slice(0, None)
                     gen_baseline_batch.non_tensor_batch["__do_sample__"] = np.zeros(len(gen_baseline_batch), dtype=bool)
                     combined_gen_batch = DataProto.concat([gen_batch_output, gen_baseline_batch])
                     num_sampled_prompts = len(gen_batch_output)

--- a/verl/trainer/ppo/ray_trainer.py
+++ b/verl/trainer/ppo/ray_trainer.py
@@ -22,7 +22,6 @@ import json
 import os
 import uuid
 from collections import defaultdict
-from copy import deepcopy
 from pprint import pprint
 from typing import Any, Optional
 

